docs/spec/namespace.md

```diff
@@ -0,0 +1,198 @@
+# Namespace
+The global namespace represents the full set of named repositories which can be
+referenced by Docker. Any scope within the namespace may refer to a subset of
+the named repos. A repository contains a set of content-addressable blobs and
+tags referencing those blobs. The namespace can be used to discover the
+location and certificate information of a repository based on DNS, HTTP
+requests and other means.
+
+A repository should always be referenced by its fully qualified name. If a
+client presents a shortened name to a user, that name should be fully expanded
+based on rules defined by the client before contacting a remote server. When
+mirroring repositories, the original name for a repository must be used.
+Changing a repository name is equivalent to copying or moving and will not be
+referenced by the original repository.
+
+## Terminology
+
+ - *Global Namespace* - The full set of referenceable names.
+ - *Name* - A fully qualified string containing both the domain and resource name
+ - *Repository* - A collection of objects under the same name within the
+namespace.
+ - *Namespace* - also *Namespace Scope* - A collection of repositories with a
+common name prefix and set of services including registry API, index, and trust
+context.
+ - *Short Name* - A name which does not contain a domain and requires
+expansion to a fully qualified name before resolving to a repository.
+
+## Format
+A name consists of two parts, a DNS host name plus a repository path. The host
+name follows the DNS rules without change. The total length of a name is 255
+characters however there is no specific limitation on DNS or path components.
+
+### Name Grammar
+```
+<name> ::= <hostname>"/"<path>
+<hostname> ::= <host-part>*["."<host-part>][":"<port>]
+<path> ::= <path-part>*["/"<path-part>]
+<host-part> ::= <regexp "[a-z]([-]?[a-z0-9])*">
```

Comment:
@yosifkit A few constraints on what is accepted are acceptable in the face of consistency and security. I am not sure that I agree that this would limit where a user can host a registry. Domains are cheap or free. Do you have specific examples of situations where people want to host images in a domain that can't be represented and they absolutely cannot change the domain?

Check out #609 for the work on the `reference` package. We are putting together a full BNF for a "Reference" which will capture all of this.

Either way, it's easier to make this more accepting later than it is to lock this down further in the future. We can make this suggested change at any time, but going back to the more restrictive format would be impossible.